Hands-On Lab: Diagnosing and Fixing Model Fit   
● Step 1: Deliberately overfit a model (e.g. a very deep decision tree) and confirm the large train-vs
validation gap.     
● Step 2: Deliberately underfit a model (e.g. an overly simple one) and confirm both scores are low.       
● Step 3: Apply regularization (or reduce complexity) to the overfit model and show the gap shrink.     
● Step 4: Document each diagnosis and fix with the score evidence in Markdown.      

At first definitions:   
Bias: is error from wrong assumptions (too simple).  
Cariance: is error from sensitivity to the specific training data (too complex).

Bias and Cariance are negativly correlated. 

Our goal is to be somewhere in the middle.

In [38]:
from sklearn.datasets import fetch_california_housing , load_breast_cancer

from sklearn.tree import DecisionTreeRegressor , DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score ,accuracy_score

Overfitting vs Underfitting (Regression) :

In [5]:
housing = fetch_california_housing()

X = housing.data
y = housing.target

X_train , X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Overfitting: 

In [24]:
DT = DecisionTreeRegressor(max_depth= None,random_state=42)
DT.fit(X_train,y_train)

train_prediction = DT.predict(X_train)
test_prediction = DT.predict(X_test)

train_r2 = r2_score(y_train, train_prediction)
test_r2 = r2_score(y_test, test_prediction)

print (f"REG Train R^2 score when 'Overfitted' : {train_r2}")
print (f"REG Test R^2 score when 'Overfitted' : {test_r2}")

REG Train R^2 score when 'Overfitted' : 1.0
REG Test R^2 score when 'Overfitted' : 0.622823312540521


A perfect 1.0 on the training data means the model perfectly memorized the price of every single house in the training set.     
 But when we ask it to predict the prices of the 20% of houses it didn't see during training (the validation set), its accuracy drops to 62%.

Underfitting:

In [25]:
DT_u = DecisionTreeRegressor(max_depth= 1,random_state=42)
DT_u.fit(X_train,y_train)

train_preiction_u = DT_u.predict(X_train)
test_prediction_u = DT_u.predict(X_test)


train_r2_u = r2_score(y_train, train_preiction_u)
test_r2_u = r2_score(y_test, test_prediction_u)


print (f"REG Train R^2 score when 'Underfitted' : {train_r2_u}")
print (f"REG Test R^2 score when 'Underfitted' : {test_r2_u}")

REG Train R^2 score when 'Underfitted' : 0.3169966266359022
REG Test R^2 score when 'Underfitted' : 0.2795112163463813


For underfitting low scores in both training and testing sets is expected since it doesn't get much to train nor to base decisions of.

To cure the high variance from overfitting or the high bias from underfiiting,  
we apply "regularization" by limiting the complexity of the tree.

Reguralization adds a penalty for model complexity,     
discouraging the model from relying too heavily on any single feature.

 In linear models it can be done using Ridge and Lasso regression.

Basically reguralization aims at reducing complexity and it can be achived in multiple ways.        
     -- in our case of DecisionTrees we can reduce the max depth(which correlates to how much the model knows about data or how much it asks).

 Regularized:

In [ ]:
DT_r = DecisionTreeRegressor(max_depth=7, random_state=42)
DT_r.fit(X_train,y_train)

train_preiction_r = DT_r.predict(X_train)
test_prediction_r = DT_r.predict(X_test)


train_r2_r = r2_score(y_train, train_preiction_r)
test_r2_r = r2_score(y_test, test_prediction_r)

print (f"REG Train R^2 score when 'Regularized' : {train_r2_r}")
print (f"REG Test R^2 score when 'Regularized' : {test_r2_r}")

REG Train R^2 score when 'Regularized' : 0.7166975587892221
REG Test R^2 score when 'Regularized' : 0.6515700686011479


REG Train R^2 score when 'Regularized' : 0.7166975587892221     
REG Test R^2 score when 'Regularized' : 0.6515700686011479      

As seen, After getting regularized  it showed a huge improvment in consistancy due to the reduction of variance.    

Train overfitted: 1.00 vs 0.71 in regularized.    
Train overfitted: 0.62 vs 0.65 in regularized.   

we can see that the gap between training and testing is significantly smaller in the regularized version.






Overfitting vs Underfitting (Classification) :

In [ ]:
cancer = load_breast_cancer()

X1 = cancer.data
y1 = (cancer.target == 0).astype(int) # maliginant = 1 , bengin = 0 

X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1 , test_size= 0.2, random_state = 42)

Overfitting:

In [42]:
DTclf = DecisionTreeClassifier(max_depth= None,random_state=42)
DTclf.fit(X1_train,y1_train)

train_prediction_clf = DTclf.predict(X1_train)
test_prediction_clf = DTclf.predict(X1_test)

train_r2_clf = accuracy_score(y1_train, train_prediction_clf)
test_r2_clf = accuracy_score(y1_test, test_prediction_clf)

print (f"CLF Train Accuracy score when 'Overfitted' : {train_r2_clf}")
print (f"CLF Test Accuracy score when 'Overfitted' : {test_r2_clf}")


CLF Train Accuracy score when 'Overfitted' : 1.0
CLF Test Accuracy score when 'Overfitted' : 0.9473684210526315


Underfitting: 

In [43]:
DTclf_u = DecisionTreeClassifier(max_depth= 1,random_state=42)
DTclf_u.fit(X1_train , y1_train)

train_preiction_clf_u = DTclf_u.predict(X1_train)
test_prediction_clf_u = DTclf_u.predict(X1_test)


train_r2_clf_u = accuracy_score(y1_train, train_preiction_clf_u)
test_r2_clf_u = accuracy_score(y1_test, test_prediction_clf_u)


print (f"CLF Train Accuracy score when 'Underfitted' : {train_r2_clf_u}")
print (f"CLF Test Accuracy score when 'Underfitted' : {test_r2_clf_u}")

CLF Train Accuracy score when 'Underfitted' : 0.9208791208791208
CLF Test Accuracy score when 'Underfitted' : 0.8947368421052632


CLF Train Accuracy score when 'Overfitted' : 1.0    
CLF Test Accuracy score when 'Overfitted' : 0.9473684210526315      

CLF Train Accuracy score when 'Underfitted' : 0.9208791208791208    
CLF Test Accuracy score when 'Underfitted' : 0.8947368421052632                                                    

Regularized:

In [44]:
DTclf_r = DecisionTreeClassifier(max_depth=3, random_state=42)

DTclf_r.fit(X1_train,y1_train)

train_preiction_clf_r = DTclf_r.predict(X1_train)
test_prediction_clf_r = DTclf_r.predict(X1_test)


train_r2_clf_r = accuracy_score(y1_train, train_preiction_clf_r)
test_r2_clf_r = accuracy_score(y1_test, test_prediction_clf_r)


print (f"CLF Train Accuracy score when 'Regularized' : {train_r2_clf_r}")
print (f"CLF Test Accuracy score when 'Regularized' : {test_r2_clf_r}")

CLF Train Accuracy score when 'Regularized' : 0.978021978021978
CLF Test Accuracy score when 'Regularized' : 0.9473684210526315


## Diagnosing and Fixing Model Fit

### Part 1: Regression (California Housing Dataset)

| Model Configuration | Training $R^2$ | Testing $R^2$ | Diagnosis |
| :--- | :--- | :--- | :--- |
| **Overfit** (`max_depth=None`) | 1.000 | 0.622 | High Variance |
| **Underfit** (`max_depth=1`) | 0.316 | 0.279 | High Bias |
| **Regularized** (`max_depth=7`) | 0.716 | 0.651 | Good Fit |

*   **Overfitting Diagnosis:** With `max_depth=None`, the model memorized the training data entirely, achieving a perfect $R^2$ score of 1.0. However, performance dropped to roughly 0.62 on the test set, creating a massive gap that is a textbook symptom of high variance.

*   **Underfitting Diagnosis:** Constraining the depth to 1 choked the model, resulting in poor scores of approximately 0.31 and 0.27 on the training and test sets, respectively. This indicates high bias because the model was too simple to capture the underlying pattern in the housing data.

*   **The Fix:** By setting `max_depth=7`, we regularized the model by reducing its complexity. The training score correctly dropped to roughly 0.71, while the test score climbed to 0.65. This shrank the train-validation gap significantly, proving the model is now generalizing well.

---

### Part 2: Classification (Breast Cancer Dataset)

| Model Configuration | Training Accuracy | Testing Accuracy | Diagnosis |
| :--- | :--- | :--- | :--- |
| **Overfit** (`max_depth=None`) | 1.000 | 0.947 | High Variance |
| **Underfit** (`max_depth=1`) | 0.921 | 0.895 | High Bias |
| **Regularized** (`max_depth=3`) | 0.978 | 0.947 | Good Fit |

*   **Overfitting Diagnosis:** Without limits, the classification tree perfectly memorized the training data, achieving a 1.0 (100%) accuracy score. However, its performance dropped to roughly 0.947 on the test set. While 94.7% is still a high score, the gap between the perfect training score and the test score indicates the model learned specific noise from the training data rather than just the general pattern.

*   **Underfitting Diagnosis:** Restricting the tree depth to 1 forced the model to be overly simple, causing both scores to drop to their lowest points (0.921 on training and 0.895 on testing). This confirms high bias, as the model lacked the complexity needed to find the best decision boundaries.

*   **The Fix:** Tuning the model to a `max_depth=3` appropriately limited the tree's complexity. The training accuracy dropped to a more realistic 0.978, closing the gap with the test score (which remained strong at 0.947). This smaller train-vs-test gap demonstrates a much healthier, more robust model fit that generalizes well to unseen data.